### ISSUE TITLE
Register callback function to run when task is completed

### ISSUE THREAD

**Issue by markg85** [cpp-taskflow/cpp-taskflow#47]
Hi,

C++ has a bit of a weird API when it comes to `std::future` and monitoring progress. It is "doable" but certainly not trivial and requires quite a bit of hackery. What i once did is starting the tasks from within a sub thread and let that thread block on the tasks, like your `wait_for_all()` function. Then, once those tasks were done the block lifted and the next line would be called to in effect notify some other class that the task is done.

I'm hoping for an api function like that in taskflow.
You now have the functions: `dispatch/silent_dispatch/wait_for_topologies/wait_for_all`
I would like to request a new addition to those, lets call it `dispatch_callback` which would be non blocking and takes one callable as argument. That callable would be called once all tasks are done processing.

The implementation could be what i described above as hackery :) But i'm guessing you have much neater options at your disposal as you're already managing the threadpool behind it.

For me, it would make this project really useful! Even though i'm merely using it as a threadpool, it simply becomes very easy and convenient to use and implement multithreading in something :)

Cheers,
Mark

**Comment by twhuang-uiuc** [comment]
Hi @markg85 I completely agreed with you on this. I will discuss with the team to add this functionality.

**Comment by markg85** [comment]
Hi @twhuang-uiuc, that's great!

Please do note that it would be quite desirable to have the callback function be executed on the same thread as where the treadpool (or taskflow) object was created. Aka, the main thread (usually).

The workaround that i described did not do that, the callback was executed from the "helper thread". This complicates matters when you want to update a GUI with the callback.

**Comment by twhuang-uiuc** [comment]
Hi @markg85 I am a bit unsure why you would like to restrict the callback be called by the owner (usually main thread). Calling `dispatch` will dispatches the task dependency graph to the threads and whoever finish the last task can immediately invoke the callback. If we want to enforce the main thread to run the callback, it can be very difficult because the main thread might be doing something else to overlap the graph execution. Does this make sense to you? 

Thanks for you comments.

**Comment by markg85** [comment]
Hi @twhuang-uiuc, I will try to explain it the best i can.

What you suggest probably works fine in non-gui environments.
Lets take a gui (like Qt) as an example.
Imagine you have a heavy task, you want to execute it in parallel and show the user a spinner (or some form of progress) while the task is running. All GUI updates would have to be done on the main thread! That's simply how GUI libraries work.

So imagine the following pseudo code:
```
QApplication app;
QMovie spinner("spinner.gif")
spinner.centerIn(app);
spinner.setVisible(false);
tf::Taskflow tf;
// Add the heady task
spinner.setVisible(true);
spinner.start();
tf.dispatch_callback([&spinner]()
{
  // This part needs to be executed on the GUI thread!
  // It won't work when called from another thread
  spinner.stop();
  spinner.setVisible(false);
});
```

Note the code is **not** actual Qt code. I made it up as i typed it. I greatly simplified how things work. The only things that matter are the start, stop and setVisible functions (those do exist). They need to be run on the GUI thread.

When the callback is executed on the same thread as where the taskflow object was created (and the GUI was) then the callback will work and in this case stop and hide the spinner.

When the TaskFlow object is on any other thread then the main thread, GUI actions won't work from it's callback.

In the case of Qt you'd then have to work with either signals/slots but that requires the callback object to live in a QObject derived class. Or by sending an event to the application event queue (that can be done from a callback and from any thread) which is then handled on the GUI thread by some more code for handling that specific event.

As you can see, not being on the GUI thread for the callback will cause some pain. That's not just with TaskFlow, it's with any thread system.

Does this help?

**Comment by clin99** [comment]
Hi @markg85 , If you want the callback to be executed by the owner thread itself, then you can directly invoke 
```
 spinner.stop();
 spinner.setVisible(false);
```
in the owner thread without dispatching a callback.  In your case, these two lines are restricted to the owner thread, which means they will be eventually (and can only be) executed by the owner thread at some time and putting them in another callback seems redundant here.  Please correct me if my understanding is wrong.

Thanks.

**Comment by markg85** [comment]
Hi @clin99, i'm not so sure about that.

What you say would only work if the hypothetical **tf.dispatch_callback** would be blocking.
Which completely defeats the point of having a non blocking call that calls the callable object when done :)

so what you suggest is basically this (expanding the above example a bit):
```
...
spinner.start();
tf.wait_for_all();
spinner.stop();
spinner.setVisible(false);
...
```
Logic like that is working now, but that blocks the GUI thread which in turn causes the spinner to not even spin (as the thread is blocked).

and if you would do this with the hypothetical **tf.dispatch_callback** (lets take silent_dispatch as an example) is this:
```
...
spinner.start();
tf.silent_dispatch();
spinner.stop();
spinner.setVisible(false);
...
```
which would result in no spinner at all as - again - the silent_dispatch is non blocking thus the next lines get executed immediately.. Not exactly what i had in mind.

**Comment by clin99** [comment]
Hi @markg85 , based on my understanding, what you want is to let the owner thread do the ``stop`` and ``setVisible(false)`` after all tasks are done, and in order to spin the spinner during task execution the owner thread should be acknowledged in a non-blocking way when all tasks are done. Is this correct?

Thanks.

**Comment by markg85** [comment]
Hi @clin99 , yes, that's correct.
Come to think of it some more, i wonder what kind of solution you have for this as it would be darn difficult without an event loop running on the main thread where you could hook in.

**Comment by clin99** [comment]
Hi @markg85 , Thanks for confirming this. Currently I can't think of any good solution for this. I agree that making the owner thread in a polling loop is a feasible solution and this does not require changing the Taskflow.

**Comment by markg85** [comment]
Hi @clin99 and @twhuang-uiuc 

Perhaps just go for the initial thought of letting the last finished task call the callback.
But please do clearly document with that callback function that it can be called from any thread, as that is quite important.
It would then be the bare groundwork on which to build further by those using it, all depending on the usecase obviously.

**Comment by twhuang-uiuc** [comment]
Hi @markg85 
I am guessing what you want is a reactor-like event-driven framework. If yes, I completely understand the importance to have the master thread control certain things. In the past I designed a reactor that have similar functionalities to enforce a callback to be invoked by the master thread - in particular, forking the process. I then use cpp-taskflow to assist the reactor in making events run in parallel. For example, the reactor detects some active events and then dispatches a task dependency graph and so on. Perhaps this can be a feasible solution to your case?  

Thank you for the useful discussion :)

**Comment by markg85** [comment]
Hi @twhuang-uiuc 
I think that would be good, but not for this framework.
It's simply out of scope.

It would probably be best for this library to provide the building blocks to implement something like that. Having a callback is the only missing building block as far as i can see.

There might be one more useful addition that would aid implementing it. imagine having multiple taskflow instances for multiple cpu intensive tasks. If an event system is used to identify that a task is done it might be nice to also be able to identify which task is done. It can be as simple as passing around the pointer of the taskflow object to the callback function. As then the event can specify that task X is done when task Y might still be running.

**Comment by markg85** [comment]
Any update on this?

**Comment by twhuang-uiuc** [comment]
Hi @markg85 , we are working on this. Like we said, we may not guarantee the callback to run on the master thread. Will this be still useful for you? 

Thanks for your comments.

**Comment by markg85** [comment]
Hi @twhuang-uiuc , This still certainly help!
As said before, it allows for easy and deep integration.
Without it that's also possible, just by doing it in the application code as opposed to in your header files.

**Comment by clin99** [comment]
Hi @markg85 ,     Now both the `dispatch` and `silent_dispatch` can accept a callback and the callback will be invoked (on one worker thread) after all tasks complete. Please refer to the [dispatch](https://github.com/cpp-taskflow/cpp-taskflow/blob/master/example/dispatch.cpp) example for how to register a callback.

Thanks.

**Comment by twhuang-uiuc** [comment]
Hi @markg85 , is it ok to close this issue?

**Comment by markg85** [comment]
@clin99 That is awesome! Thank you very much!
@twhuang-uiuc yup, will close it with this comment :)

**Comment by limingchina** [comment]
> Hi @markg85 , now both the `dispatch` and `silent_dispatch` can accept a callback and the callback will be invoked (on one worker thread) after all tasks complete. Please refer to the [dispatch](https://github.com/cpp-taskflow/cpp-taskflow/blob/master/example/dispatch.cpp) example for how to register a callback.
> 
> Thanks.

The dispatch example doesn't exist in the lastest master. Is the functionality replaced with observer?
